[lexical] Feature: registerMutationListener should initialize its existing nodes

etrepum opened the pull request:
## Description

When `registerMutationListener` is called, if there are any existing nodes with that type, then the listener should immediately be called once with any existing nodes in the DOM from the previously reconciled state.

This makes it easier to implement mutation listeners correctly, especially for plug-ins that may be loaded lazily, and mirrors the behavior for `registerRootListener` and `registerNodeTransform` where it is expected that the code will be run for existing content (though there are other types listeners that do not have this behavior, I think it really makes the most sense for mutation listeners to behave this way). Some existing mutation listeners either did not support lazy loading, or did this incorrectly (e.g. by checking for an exact node type match instead of considering node replacement as in #6349).

This new "synthetic" mutation can be detected from the listener by checking for the `registerMutationListener` update tag, and it can be skipped entirely with the new `{skipInitialization: true}` option argument (mutation listeners registered in this way will have the same semantics as before this PR).

Closes #6356
Closes #6349

## Test plan

Unit tests are included. Only one existing unit test needed to be changed to use the new `{skipInitialization: true}` option.

I did have to fix an egregious react rules violation in [CodeActionMenuPlugin](https://github.com/facebook/lexical/pull/6357/commits/c73a70a6f526f4be6cee43081e3eb2560d3bffef) that caused an infinite rendering loop with this change, the registration needed to be wrapped in `useEffect`. The previous behavior was basically a memory & performance leak, fortunately only in playground code.

zurfyx reviewed (commented):
Thanks Bob! This is intentional:
1. Conceptually Nodes aren't "created" when you register the function.
2. The listeners weren't meant to be called immediately upon registration (the Root listener was meant to be a one-off).

Note that this is also a dangerous breaking change, the API is backward compatible but the behavior is not. Internally, we commonly pair this function with a Map to extend the behavior of the Node independently and these now repeated events can cause trouble with the initialization.

https://github.com/facebook/lexical/pull/4099 is a potential option to mitigate the issue. I think it is reasonable to have this build-in via a property (don't have strong opinions on 2.) but at the same time I'd advocate for a more flat API instead of yet another configuration object.

Would like to hear what other think about this as well.

etrepum commented:
`$nodesOfType` is not good for this use case because it only considers exact type, not node replacement by subclass, so it creates a different initial set than this version of `registerMutationListener`. If `$nodesOfType` worked I wouldn't really have been motivated to fix `registerMutationListener` this way.

etrepum commented:
Other than the number of calls, the breaking API change is unlikely to cause problems because it will just emit an initial created event where previously you had to do a separate read to discover them and this was only implemented correctly one time in the monorepo (it was implemented wrong in two places, and not implemented at all in several other places). Implementing `registerMutationListener` in this way removes the need for #4099 altogether, making the API simpler, at the cost of a mostly harmless breaking API change.

Alternatively the new behavior could be added with an option, or a separate function/method, but I would prefer the breaking change because it makes it much easier to write correct code and will probably even fix a few latent bugs.

fantactuka commented:
> 1. Conceptually Nodes aren't "created" when you register the function.
> 2. The listeners weren't meant to be called immediately upon registration (the Root listener was meant to be a one-off).

It depends on how editor's state is initialized. If `initialEditorState` is async then mutation listener will be initialized before it's executed and called for initial notes. It's likely less common use case though

> Note that this is also a dangerous breaking change, the API is backward compatible but the behavior is not. Internally, we commonly pair this function with a Map to extend the behavior of the Node independently and these now repeated events can cause trouble with the initialization.

+1, I did check registerMutationListener + $nodesOfType (commonly used to handle initial nodes) internally and it relies on nodeID which should avoid double initialization, but there could be usages that just go and mutate some DOM for 'created' mutation without tracking nodeID at all

fantactuka commented:
Re: `registerMutationListener` vs `registerNodesOfType`, I do agree with Bob that keeping simpler API is better here. Can think about rollout plan to simplify migration out of nodesOfType+registerMutationListener (e.g., making it opt-in for patch version bumb, and opt-out for minor)

etrepum commented:
An opt-in migration path sounds fine to me, we can use the existing option argument for that and explicitly use `{skipInitialization: false}` to get the new behavior now. The only place I was motivated to use `{skipInitialization: true}` in the monorepo was the TOC plugin (the one correct init-before-register implementation) and only because it is tracking state for multiple mutation listeners in one place so it would've been awkward to initialize it efficiently with the pair of initial calls.

zurfyx commented:
> `$nodesOfType` is not good for this use case because it only considers exact type, not node replacement by subclass, so it creates a different initial set than this version of `registerMutationListener`. If `$nodesOfType` worked I wouldn't really have been motivated to fix `registerMutationListener` this way.

This is not intentional, this function should look into replacements like we do elsewhere.

> Re: `registerMutationListener` vs `registerNodesOfType`, I do agree with Bob that keeping simpler API is better here. Can think about rollout plan to simplify migration out of nodesOfType+registerMutationListener (e.g., making it opt-in for patch version bumb, and opt-out for minor (or v1 if we do it soon enough))

@fantactuka Can you expand on what a simpler API means for you? In either case, it would be a one-liner. 

The concern I raised about the current implementation is that previously created nodes are also reported as `created`, and the (not-so-important) fact that we need to introduce another object to carry more conditionals once more.

etrepum commented:
The `registerNodesOfType` utility from #4099 is useful, but for different use cases (sometimes you do care about update or destroy, rather then tracking a single set). In fact, it would be even simpler to implement `registerNodesOfTypeListener` with this version of `registerMutationListener` because you can get rid of the separate initialization code.

```ts
export function registerNodesOfTypeListener(
  editor: LexicalEditor,
  klass: Klass<LexicalNode>,
  listener: (nodes: Set<NodeKey>) => void,
  fireImmediately: void | boolean = false,
): () => void {
  let nodes = new Set<NodeKey>();
  return editor.registerMutationListener(klass, (mutations, {updateTags}) => {
    let newNodes: Set<NodeKey> | null = null;
    for (const [key, mutation] of mutations) {
      const created = mutation === 'created';
      const destroyed = mutation === 'destroyed';
      if (created || destroyed) {
        if (newNodes === null) {
          newNodes = new Set(nodes);
        }
        if (created) {
          newNodes.add(key);
        }
        if (destroyed) {
          newNodes.delete(key);
        }
      }
    }
    if (newNodes !== null) {
      nodes = newNodes;
      if (fireImmediately || !updateTags.has('registerMutationListener')) {
        listener(nodes);
      }
    }
  }, {skipInitialization: false});
}
```

Other than a potential complexity/performance trade-off of tracking previously created nodes separately (which is opt-in here), what are the use cases for having a mutation listener that ignores initial creates? I think in most cases the existing code is just wrong (doesn't support lazy loading at all or doesn't correctly track existing nodes), and if it is right then it is probably already has "defenses" to work around the fact that it is not seeing a complete picture of the DOM nodes (e.g. tracking nodes in a separate map so that an update can be considered to be a create) so the worst case is usually just doing extra work (which would've happened in the TOC plug-in if I didn't opt-in to `skipInitialization`) or breaking a very sensitive test (e.g. tracking the number of calls). There was one case where it exposed a bug (`CodeActionsMenu` calling register outside of an effect with no clean-up!).

etrepum commented:
FWIW I don't think `$nodesOfType` can be fixed without taking an `editor` argument or depending on an `activeEditor` from `editor.update` (or #6347's `editor.read`) since the transform information isn't available on the `EditorState`.

etrepum commented:
I've added an internal cache (`getCachedTypeToNodeMap`) for the expensive calculation so that unnecessary work is not repeated when multiple mutation listeners are added in sequence without a reconciliation between them. This cache is also used by `$nodesOfType` when called with a read-only `activeEditorState`.

I've also changed the default to `skipInitialization: true` for backwards compatibility with notes in the documentation that this default will change in 0.17.0.

etrepum commented on packages/lexical-playground/src/plugins/CodeActionMenuPlugin/index.tsx:
```diff
@@ -109,26 +109,32 @@ function CodeActionMenuContainer({
+        });
+        setShouldListenMouseMove(codeSetRef.current.size > 0);
+      },
+      {skipInitialization: false},
```
CodeActionMenuPlugin did not previously handle initialization correctly

etrepum commented on packages/lexical-code/src/CodeHighlighter.ts:
```diff
@@ -818,18 +818,22 @@ export function registerCodeHighlighting(
-    }),
+        });
+      },
+      {skipInitialization: false},
```
CodeHighlighter did not previously handle initialization correctly

ivailop7 replied:
nice catch

etrepum commented on packages/lexical-playground/src/plugins/TableActionMenuPlugin/index.tsx:
```diff
@@ -183,17 +183,21 @@ function TableActionMenu({
+          setBackgroundColor(currentCellBackgroundColor(editor) || '');
+        }
+      },
+      {skipInitialization: true},
```
TableActionMenuPlugin did handle initialization correctly (it only listens for updates)

etrepum commented on packages/lexical-react/src/LexicalAutoEmbedPlugin.tsx:
```diff
@@ -105,24 +105,23 @@ export function LexicalAutoEmbedPlugin<TEmbedConfig extends EmbedConfig>({
-    (key: NodeKey) => {
-      editor.getEditorState().read(async function () {
+    async (key: NodeKey) => {
+      const url = editor.getEditorState().read(function () {
```
This code was very sketchy previously, and only worked at all because it used `linkNode.__url`. We should probably have a lint that tells you not to use async for read or update

etrepum commented on packages/lexical-react/src/LexicalAutoEmbedPlugin.tsx:
```diff
@@ -146,7 +145,9 @@ export function LexicalAutoEmbedPlugin<TEmbedConfig extends EmbedConfig>({
       ...[LinkNode, AutoLinkNode].map((Klass) =>
-        editor.registerMutationListener(Klass, (...args) => listener(...args)),
+        editor.registerMutationListener(Klass, (...args) => listener(...args), {
+          skipInitialization: true,
```
This implementation is not compatible with `skipInitialization: false` because it resets if it sees `key === nodeKey` which would fire with `'created'`, could also be fixed by checking for `mutation !== 'created' && key === nodeKey`

etrepum commented on packages/lexical-react/src/LexicalTableOfContentsPlugin.tsx:
```diff
@@ -230,6 +230,8 @@ export function TableOfContentsPlugin({children}: Props): JSX.Element {
         });
       },
+      // Initialization is handled separately
+      {skipInitialization: true},
```
LexicalTableOfContentsPlugin already had an efficient and correct initialization (though this wouldn't break it)

etrepum commented on packages/lexical-react/src/LexicalTablePlugin.ts:
```diff
@@ -129,17 +128,6 @@ export function TablePlugin({
-    // Plugins might be loaded _after_ initial content is set, hence existing table nodes
-    // won't be initialized from mutation[create] listener. Instead doing it here,
-    editor.getEditorState().read(() => {
-      const tableNodes = $nodesOfType(TableNode);
```
This was not correct because `$nodesOfType` can't consider node replacement without depending on the editor

etrepum commented on packages/lexical-website/docs/concepts/dom-events.md:
````diff
@@ -30,20 +30,20 @@ This can be a simple, efficient way to handle some use cases, since it's not nec
+In some cases, it may be better to attach an event handler directly to the underlying DOM node of each specific node. With this approach, you generally don't need to filter the event target in the handler, which can make it a bit simpler. It will also guarantee that your handler isn't running for events that you don't care about. This approach is implemented via a [Mutation Listener](https://lexi ...
 
 ```js
+const registeredElements: WeakSet<HTMLElement> = new WeakSet();
````
Just fixing the correctness and formatting of this relevant example, the state needs to be outside of the listener implementation!

etrepum commented on packages/lexical/src/LexicalEditor.ts:
```diff
@@ -825,15 +838,43 @@ export class LexicalEditor {
+    ).klass;
+    const mutations = this._listeners.mutation;
+    mutations.set(listener, klassToMutate);
+    const skipInitialization = options && options.skipInitialization;
+    if (
+      !(skipInitialization === undefined
+        ? DEFAULT_SKIP_INITIALIZATION
+        : skipInitialization)
+    ) {
```
Would be `const skipInitialization = options?.skipInitialization ?? DEFAULT_SKIP_INITIALIZATION;` if we could use null coalescing

StyleT reviewed (approved):
LGTM!

etrepum commented:
I updated the branch and the tests are passing again (I guess main was broken earlier this week)

ivailop7 commented:
> I updated the branch and the tests are passing again (I guess main was broken earlier this week)

LGTM, could you check the new TableHoverActions plugin I added as well for rules violation as well, I copied a lot of from the CodeActions plugin and I'm on Mobile right now.

etrepum commented:
@ivailop7 good catch, the new `TableHoverActions` plugin did require `{skipInitialization: false}` for correct lazy initialization.

ivailop7 commented:
> @ivailop7 good catch, the new `TableHoverActions` plugin did require `{skipInitialization: false}` for correct lazy initialization.

Thanks!

ivailop7 commented:
Leaving to @zurfyx to merge

zurfyx commented on packages/lexical/src/LexicalEditor.ts:
```diff
@@ -842,49 +883,51 @@ export class LexicalEditor {
+    for (const k of nodeMap.keys()) {
+      nodeMutationMap.set(k, 'created');
+    }
+    if (nodeMutationMap.size > 0) {
```
For the sake of predictability I wonder if we should always call it regardless. This is sort of inline with my [previous comment](https://github.com/facebook/lexical/pull/6357#issuecomment-2200445030) but less radical, I understand that libraries like Rxjs follow similar patterns.

⭐️ This predicatibility also gives users more flexibility and the option flag becomes more of a convenience rather than a must-have.

etrepum replied:
I don't have a strong preference one way or the other, correct code won't notice a difference. The current zero check is "more" backwards compatible in that the behavior only changes if the mutation is registered after the initial reconciliation, so I think more tests will end up having to change a bit (the ones based on call count rather than meaningful substance). Since we ended up with an option and a deprecation phase I think either is fine. Anyone else have an opinion about this?

zurfyx commented on packages/lexical-website/docs/concepts/listeners.md:
```diff
@@ -81,15 +81,21 @@ Get notified when a specific type of Lexical node has been mutated. There are th
 const removeMutationListener = editor.registerMutationListener(
   MyCustomNode,
-  (mutatedNodes) => {
+  (mutatedNodes, { updateTags, dirtyLeaves, prevEditorState }) => {
```
Is this correct?

etrepum replied:
Yes, as far as I can tell
```ts
export type MutationListener = (
  nodes: Map<NodeKey, NodeMutation>,
  payload: {
    updateTags: Set<string>;
    dirtyLeaves: Set<string>;
    prevEditorState: EditorState;
  },
) => void;
```

zurfyx commented on packages/lexical/src/LexicalEditor.ts:
```diff
@@ -211,6 +212,18 @@ export type MutatedNodes = Map<Klass<LexicalNode>, Map<NodeKey, NodeMutation>>;
+   * The default is currently true for backwards compatibility with <= 0.16.1
+   * but this default is expected to change to false in 0.17.0.
+   */
+  skipInitialization?: boolean;
```
Is this the best term for users? Initialization describes the work done behind the users, for users it's merely skipping the first one. Do we even need? ⭐️

etrepum replied:
We need an option because on the call it was discussed that we need the default behavior to be exactly the current behavior. When the default changes, it should be very rare to have to specify this option, which is why it should have a name that implies that the default is false. I don't have a strong preference as to whether it's called "skipInitialization" or "doNotSendCreateEventForExistingDOM" or whatever.

zurfyx commented on packages/lexical/src/LexicalUtils.ts:
```diff
@@ -1118,16 +1118,21 @@ export function setMutatedNode(
       node instanceof klass &&
       node.__type === klassType &&
-      (readOnly || node.isAttached())
+      node.isAttached()
```
I think this is correct. I believe I added the `readOnly` flag before the GC fix

etrepum replied:
`readOnly` can never happen here anymore because there's an early exit for that case above

```ts
  if (editorState._readOnly) {
    const nodes = getCachedTypeToNodeMap(editorState).get(klassType) as
      | undefined
      | Map<string, T>;
    return nodes ? [...nodes.values()] : [];
  }
```

zurfyx commented on packages/lexical/src/LexicalUtils.ts:
```diff
@@ -1691,3 +1696,34 @@ export function $getAncestor<NodeType extends LexicalNode = LexicalNode>(
 }
+
+/** @internal */
+export type TypeTodNodeMap = Map<string, NodeMap>;
```
Nit, typo
```suggestion
export type TypedToNodeMap = Map<string, NodeMap>;
```

etrepum replied:
Good catch! Looks like I didn't finish a refactor from one naming convention to another *and* had a typo in there.

zurfyx reviewed (approved):
Stamping, sorry for the back and forth, let's maybe take some time next session to discuss the nits that are relevant for the version bump. Merging as soon as the tests pass since @ivailop7 and @StyleT have already given their thumb up
